format.md

```diff
@@ -0,0 +1,304 @@
+- Full data type system of Arrow is available for attribute data.
+
+More specifically, the Arrow geometry specification stores the raw coordinates
+values in contiguous arrays with enough metadata (offsets) to reconstruct or
+interpret as actual geometries.
+
+[FlatGeoBuf](https://flatgeobuf.org/) is similar on various aspects, but is
+record-oriented, while Arrow is column-oriented.
+
+## Format
+
+The terminology for array types in this section is based on the
+[Arrow Columnar Format specification](https://arrow.apache.org/docs/format/Columnar.html).
+
+GeoArrow proposes a packed columnar data format for the fundamental geometry
+types, using packed coordinate and offset arrays to define geometry objects.
+
+The inner level is always a FixedSizeList array storing the interleaved
+coordinate values (`[x, y, x, y, ...]`, physically stored as a single array
+of length `n_dim * n_coords`). For any geometry type except Point, this inner
+level is nested in one or multiple
+[variable sized list](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout)
+arrays. In practice, this means we have additional arrays storing the offsets
+that denote where a new geometry, a new geometry part, or a new polygon ring
+starts.
+
+**Point**
+
+* `FixedSizeList<double>[n_dim]`
+
+For an array of Point geometries, only a single array is used of interleaved
+coordinates. `n_dim` can be 2, 3, or 4 depending on the dimensionality of the
+geometries, and the field name of the list should be "xy", "xyz" or "xyzm",
+reflecting the dimensionality.
+
+**LineString**
+
+* `List<FixedSizeList<double>[n_dim]>`
+
+An array of LineStrings is represented as a nested list array with one
+level of outer nesting: each element of the array (LineString) is a
+list of xy vertices. The child name of the outer list should be "vertices".
+
+**Polygon**
+
+* `List<List<FixedSizeList<double>[n_dim]>>`
+
+An array of Polygons is represented as a nested list array with two levels of
+outer nesting: each element of the array (Polygon) is a list of rings (the
+first ring is the exterior ring, optional subsequent rings are interior
+rings), and each ring is a list of xy vertices. The child name of the outer
+list should be "rings"; the child name of the inner list should be "vertices".
+
+**MultiPoint**
+
+* `List<FixedSizeList<double>[n_dim]>`
+
+An array of MultiPoints is represented as a nested list array, where each outer
+list is a single MultiPoint (i.e. a list of xy coordinates). The child name of
+the outer `List` should be "points".
+
+**MultiLineString**
+
+* `List<List<FixedSizeList<double>[n_dim]>>`
+
+An array of MultiLineStrings is represented as a nested list array with two
+levels of outer nesting: each element of the array (MultiLineString) is a
+list of LineStrings, which consist itself of a list xy vertices (see above).
+The child name of the outer list should be "linestrings"; the child name of
+the inner list should be "vertices".
```

Comment:
In https://www.github.com/rapidsai/cuspatial I implemented LineStrings and MultiLineStrings in the same buffer - essentially LineStrings are just MultiLineStrings with length 1. Implementation-wise I use a bracketed-offsets buffer that might be inferior to simply incrementing the buffer by 1 for each LineString. https://github.com/rapidsai/cuspatial/blob/03d55a25c5d9f732327fcf47c98b3032904e0e0e/python/cuspatial/cuspatial/geometry/geoarrowbuffers.py#L83-L106

The reasoning behind this is that MultiLineStrings and LineStrings will frequently be used in the same algorithm, so packing them together makes optimizing those algorithms much easier.

How do you feel about combining them in your official spec here?

Reply:
Copy @harrism who is asking about the official spec to-date.